### PR TITLE
Add ability to perform CSRF token check on multi-part form post

### PR DESCRIPTION
**Pull request by dawidsawa**
This is related to change being worked in NXRM, as it seems the only way to perform the CSRF token check for ExtJS file upload forms is here, directly in `DirectJNgineServlet`.

### CODE REVIEW

File: directjngine/src/main/java/com/softwarementors/extjs/djn/servlet/DirectJNgineServlet.java [hunk 1]
```diff
@@ -78,9 +81,10 @@
   /*********************************************************  
    * GlobalParameters and configuration
    *********************************************************/
+  private static final String ANTI_CSRF_TOKEN_NAME = "NX-ANTI-CSRF-TOKEN";
```
Comment (mpiggott) [inline directjngine/src/main/java/com/softwarementors/extjs/djn/servlet/DirectJNgineServlet.java]: I wonder if this ought to be passed in via the constructor

File: directjngine/src/main/java/com/softwarementors/extjs/djn/servlet/DirectJNgineServlet.java [hunk 2]
```diff
@@ -657,4 +676,31 @@ private void processUploadFormPost(HttpServletRequest request, HttpServletRespon
...
+          return cookie.getValue();
+        }
+      }
+    }
+    return null;
+  }
+
+  private String getAntiCsrfTokenField(final List<FileItem> fileItems) {
+    for (FileItem item : fileItems) {
+      if (item.isFormField() && ANTI_CSRF_TOKEN_NAME.equals(item.getFieldName())) {
```
Comment (mpiggott) [inline directjngine/src/main/java/com/softwarementors/extjs/djn/servlet/DirectJNgineServlet.java]: All fields become `FileItem`s? Seems odd.

Comment (dawidsawa) [reply directjngine/src/main/java/com/softwarementors/extjs/djn/servlet/DirectJNgineServlet.java]: That's correct. They are later separated into files and fields [here](https://github.com/sonatype/directjngine/blob/fb7e513b4d02eaa0cbc15e237cf1f935188cf895/directjngine/src/main/java/com/softwarementors/extjs/djn/router/processor/standard/form/upload/UploadFormPostRequestProcessor.java#L62)

File: directjngine/src/main/java/com/softwarementors/extjs/djn/servlet/DirectJNgineServlet.java [hunk 3]
```diff
@@ -657,4 +676,31 @@ private void processUploadFormPost(HttpServletRequest request, HttpServletRespon
...
+        if (ANTI_CSRF_TOKEN_NAME.equals(cookie.getName())) {
+          return cookie.getValue();
+        }
+      }
+    }
+    return null;
+  }
+
+  private String getAntiCsrfTokenField(final List<FileItem> fileItems) {
+    for (FileItem item : fileItems) {
```
Comment (Blacktiger) [inline directjngine/src/main/java/com/softwarementors/extjs/djn/servlet/DirectJNgineServlet.java]: Is Java 8 available in this project? We could use a stream instead.

```
return fileItems.stream()
  .filter(FileItem::isFormField)
  .filter(item -> ANTI_CSRF_TOKEN_NAME.equals(item.getFieldName()))
  .findAny().orElse(null);
```

Comment (dawidsawa) [reply directjngine/src/main/java/com/softwarementors/extjs/djn/servlet/DirectJNgineServlet.java]: This project is running on Java 1.5 unfortunately...

File: directjngine/src/main/java/com/softwarementors/extjs/djn/servlet/DirectJNgineServlet.java [hunk 4]
```diff
@@ -657,4 +676,31 @@ private void processUploadFormPost(HttpServletRequest request, HttpServletRespon
...
 
+  private boolean isValidAntiCsrfTokenMatching(final HttpServletRequest request, final List<FileItem> fileItems) {
+    String cookie = getAntiCsrfTokenCookie(request);
+    String form = getAntiCsrfTokenField(fileItems);
+
+    return !antiCsrfTokenEnabled || (cookie != null && cookie.equals(form));
+  }
+
+  private String getAntiCsrfTokenCookie(final HttpServletRequest request) {
+    Cookie[] cookies = request.getCookies();
```
[the remark this answers is not in the document]
Comment (dawidsawa) [reply directjngine/src/main/java/com/softwarementors/extjs/djn/servlet/DirectJNgineServlet.java]: This project is running on Java 1.5 unfortunately...

**Pull request merged**